mlflow/langchain/_langchain_autolog.py

```diff
@@ -180,6 +228,682 @@ def _chain_with_retriever(model):
+            self.metrics["starts"] += 1
+
+        @override
+        def on_chain_end(
+            self,
+            outputs: Dict[str, Any],
+            *,
+            run_id: UUID,
+            inputs: Optional[Union[Dict[str, Any], Any]] = None,
+            **kwargs: Any,
+        ):
+            """Run when chain ends running."""
+            chain_span = self._get_span_by_run_id(run_id)
+            chain_span.add_event(
+                SpanEvent("end", timestamp=int(datetime.now(timezone.utc).timestamp()))
+            )
+            if inputs:
+                chain_span.set_inputs(self._get_chain_inputs(inputs))
+            self._end_span(chain_span, outputs=outputs)
+            self._log_artifact(
+                {"outputs": self._format_data_to_string(outputs), "kwargs": kwargs},
+                f"chain_end_{chain_span.span_id}",
+            )
+            self.metrics["step"] += 1
+            self.metrics["chain_ends"] += 1
+            self.metrics["ends"] += 1
+
+        @override
+        def on_chain_error(
+            self,
+            error: BaseException,
+            *,
+            inputs: Optional[Union[Dict[str, Any], Any]] = None,
+            run_id: UUID,
+            **kwargs: Any,
+        ):
+            """Run when chain errors."""
+            chain_span = self._get_span_by_run_id(run_id)
+            chain_span.add_event(
+                SpanEvent(
+                    "error",
+                    timestamp=int(datetime.now(timezone.utc).timestamp()),
+                    attributes={"error": str(error)},
+                )
+            )
+            if inputs:
+                chain_span.set_inputs(self._get_chain_inputs(inputs))
+            error_message = self._get_stacktrace(error)
+            self._end_span(chain_span, status=SpanStatus(TraceStatus.ERROR, error_message))
+            self.metrics["step"] += 1
+            self.metrics["errors"] += 1
+
+        @override
+        def on_tool_start(
+            self,
+            serialized: Dict[str, Any],
+            input_str: str,
+            *,
+            run_id: UUID,
+            tags: Optional[List[str]] = None,
+            parent_run_id: Optional[UUID] = None,
+            metadata: Optional[Dict[str, Any]] = None,
+            name: Optional[str] = None,
+            inputs: Optional[Dict[str, Any]] = None,
+            **kwargs: Any,
+        ):
+            """Start span for a tool run."""
+            if metadata:
+                kwargs.update({"metadata": metadata})
+            tool_span = self._start_span(
+                span_name=name or "tool",
+                parent_run_id=parent_run_id,
+                span_type=SpanType.TOOL,
+                run_id=run_id,
+                inputs={"input_str": input_str},
+                attributes=kwargs,
+            )
+            tool_span.add_event(
+                SpanEvent("start", timestamp=int(datetime.now(timezone.utc).timestamp()))
```

Comment:
```suggestion
                SpanEvent("start")
```

can we remove `timestamp`? `SpanEvent` has the default for `timestamp`.

Reply:
@B-Step62 what would happen if we accidentally pass a timestamp in seconds like this?

Reply:
Removed. And yes we should check if the timestamp is correctly microseconds.. I hate the conversion though 😢